### PR TITLE
Make configure_file() great again

### PR DESCRIPTION
**Pull request by nirbheek**
Needed for https://github.com/mesonbuild/meson/issues/1380.

Lots of changes. Just, like, look at the commits man. *tired face*

### CODE REVIEW

File: run_unittests.py [hunk 1]
```diff
@@ -174,6 +174,157 @@ def test_commonpath(self):
...
+        # Check dictionary
+        self.assertEqual(ret, d)
+        # Check substitutions
+        cmd = ['some', 'ordinary', 'strings']
+        self.assertEqual(substfunc(cmd, d), cmd)
+        cmd = ['@INPUT@.out', 'ordinary', 'strings']
+        self.assertEqual(substfunc(cmd, d), [inputs[0] + '.out'] + cmd[1:])
+        cmd = ['@INPUT0@.out', '@PLAINNAME@.ok', 'strings']
+        self.assertEqual(substfunc(cmd, d),
+                        [inputs[0] + '.out'] + [d['@PLAINNAME@'] + '.ok'] + cmd[2:])
```
Comment (ignatenkobrain) [inline run_unittests.py]: __[flake8]__
* [E128] continuation line under-indented for visual indent

File: run_unittests.py [hunk 2]
```diff
@@ -174,6 +174,157 @@ def test_commonpath(self):
...
+        self.assertEqual(ret, d)
+        # Check substitutions
+        cmd = ['some', 'ordinary', 'strings']
+        self.assertEqual(substfunc(cmd, d), cmd)
+        cmd = ['@INPUT@.out', '@OUTPUT@', 'strings']
+        self.assertEqual(substfunc(cmd, d),
+                         [inputs[0] + '.out'] + outputs + cmd[2:])
+        cmd = ['@INPUT0@.out', '@PLAINNAME@.ok', '@OUTPUT0@']
+        self.assertEqual(substfunc(cmd, d),
+                        [inputs[0] + '.out', d['@PLAINNAME@'] + '.ok'] + outputs)
```
Comment (ignatenkobrain) [inline run_unittests.py]: __[flake8]__
* [E128] continuation line under-indented for visual indent

File: mesonbuild/mesonlib.py [hunk 3]
```diff
@@ -521,3 +521,154 @@ def commonpath(paths):
...
+    if inputs:
+        # We want to substitute all the inputs.
+        values['@INPUT@'] = inputs
+        for (ii, vv) in enumerate(inputs):
+            # Write out @INPUT0@, @INPUT1@, ...
+            values['@INPUT{}@'.format(ii)] = vv
+        if len(inputs) == 1:
+            # Just one value, substitute @PLAINNAME@ and @BASENAME@
+            values['@PLAINNAME@'] = plain = os.path.split(inputs[0])[1]
+            values['@BASENAME@'] = base = os.path.splitext(plain)[0]
```
Comment (ignatenkobrain) [inline mesonbuild/mesonlib.py]: __[flake8]__
* [F841] local variable 'base' is assigned to but never used

File: mesonbuild/modules/gnome.py [hunk 4]
```diff
@@ -108,8 +108,15 @@ def compile_resources(self, state, args, kwargs):
             ifile = os.path.join(ifile.subdir, ifile.fname)
         elif isinstance(ifile, str):
             ifile = os.path.join(state.subdir, ifile)
+        elif isinstannce(ifile, (interpreter.CustomTargetHolder,
+                                 interpreter.GeneratedObjectsHolder)):
```
Comment (jpakkane) [inline mesonbuild/modules/gnome.py]: An extra 'n' here.

Comment (nirbheek) [inline mesonbuild/modules/gnome.py]: Doh!

**Pull request merged**